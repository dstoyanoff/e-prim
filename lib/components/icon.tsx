import { FC } from "react";
import { Box, BoxProps } from "./box";

export const DEFAULT_ICON_SIZE = 24;

export type IconProps = BoxProps<"svg"> & {
  /**
   * Overall size of the icon (square)
   * @default @see {@link DEFAULT_ICON_SIZE}
   */
  size?: number;

  /**
   * Allows overriding the viewBox size of the SVG for non-standard icons
   * @default @see {@link DEFAULT_ICON_SIZE}
   */
  viewBoxSize?: number;
};

export const BaseIcon: FC<IconProps> = ({
  size = DEFAULT_ICON_SIZE,
  color,
  viewBoxSize = DEFAULT_ICON_SIZE,
  ...props
}) => (
  <Box
    as="svg"
    width={size}
    height={size}
    viewBox={`0 0 ${viewBoxSize} ${viewBoxSize}`}
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
    color={color}
    label="icon"
    {...props}
  />
);
