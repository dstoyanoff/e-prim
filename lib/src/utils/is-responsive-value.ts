import { ResponsiveValue } from "../system/types";
import { TBreakpoint } from "../theme";

/**
 * Check if the value is an object that contains either breakpoint
 */
export const isResponsiveValue = <T>(value: ResponsiveValue<T>, breakpoints: TBreakpoint) => {
  if (typeof value !== "object") {
    return false;
  }

  return Object.keys(breakpoints).some(breakpoint => value && breakpoint in value);
};
