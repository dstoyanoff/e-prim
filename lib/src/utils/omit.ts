export const omit = <T extends object>(obj: T, ...keys: (keyof T)[]) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    if (!keys.includes(key as keyof T)) {
      acc[key] = value;
    }

    return acc;
  }, {} as Record<string, unknown>) as T;
};
