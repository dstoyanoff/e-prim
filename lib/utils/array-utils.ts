/**
 * Takes either a single element or an array and always returns an array of one or many elements
 * @param items input items of generic type
 */
export const singleOrMulti = <T>(items: T | T[]): T[] => (Array.isArray(items) ? items : [items]);

type ArrayLengthMutationKeys = "splice" | "push" | "pop" | "shift" | "unshift";

export type FixedLengthArray<T, L extends number, TObj = [T, ...Array<T>]> = Pick<
  TObj,
  Exclude<keyof TObj, ArrayLengthMutationKeys>
> & {
  readonly length: L;
  [I: number]: T;
  [Symbol.iterator]: () => IterableIterator<T>;
};
