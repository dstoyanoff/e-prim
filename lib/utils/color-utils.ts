/**
 * Converts a decimal percent to hex
 * @param p percent to convert. Should be an integer between 0 and 100, other values are bounded to that range.
 * Decimal values are rounded to the nearest integer
 */
export const percentToHex = (p: number) => {
  const percent = Math.max(0, Math.min(100, p)); // bound percent from 0 to 100
  const intValue = Math.round((percent / 100) * 255); // map percent to nearest integer (0 - 255)
  const hexValue = intValue.toString(16); // get hexadecimal representation

  return hexValue.padStart(2, "0"); // format with leading 0
};
