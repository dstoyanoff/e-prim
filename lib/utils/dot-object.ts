export type PathsToStringProps<T, TStopType = string> = T extends TStopType
  ? []
  : {
      [K in Extract<keyof T, string | number>]: [
        K,
        ...PathsToStringProps<T[K], TStopType>
      ];
    }[Extract<keyof T, string | number>];

export type Join<
  T extends (string | number)[],
  D extends string = "."
> = T extends []
  ? never
  : T extends [infer F]
  ? F
  : T extends [infer F, ...infer R]
  ? F extends string | number
    ? `${F}${D}${Join<Extract<R, (string | number)[]>, D>}`
    : never
  : string | number;

/**
 * Given a color key (dot-based deep path), resolves the actual color value
 * @param key dot-based path
 * @param object reference to the main object that should be used to find the value
 */
export const getValueFromKey = <
  TValue,
  TKey extends string = string,
  TObject = {}
>(
  key: TKey | undefined,
  object: TObject
): TValue | undefined => {
  if (!key) {
    return undefined;
  }

  const paths = key.split(".");

  return paths.reduce((accumulator, path) => {
    return accumulator ? accumulator[path] : (object as any)[path];
  }, undefined);
};
