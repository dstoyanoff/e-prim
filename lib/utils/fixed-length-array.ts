type ArrayLengthMutationKeys = "splice" | "push" | "pop" | "shift" | "unshift";

export type FixedLengthArray<T, L extends number, TObj = [T, ...Array<T>]> = Pick<
  TObj,
  Exclude<keyof TObj, ArrayLengthMutationKeys>
> & {
  readonly length: L;
  [I: number]: T;
  [Symbol.iterator]: () => IterableIterator<T>;
};
