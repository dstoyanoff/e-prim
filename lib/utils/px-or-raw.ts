export const pxOrRaw = (value: string | number, numberTransform: (value: number) => number = v => v) =>
  typeof value === "number" ? `${numberTransform(value)}px` : value;
