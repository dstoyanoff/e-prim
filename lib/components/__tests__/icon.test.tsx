/**
 * @jest-environment jsdom
 */

import { FC } from "react";
import { render } from "@testing-library/react";
import { wrapper } from "@/utils/mock-theme";
import { BaseIcon, IconProps } from "../icon";

const ArrowIcon: FC<IconProps> = props => (
  <BaseIcon {...props}>
    <path d="M4 12H20M20 12L12 4M20 12L12 20" stroke="currentColor" strokeLinecap="round" strokeLinejoin="round" />
  </BaseIcon>
);

describe("icon", () => {
  test("should render icon", () =>
    expect(
      render(<ArrowIcon color="neutral.4" size={32} />, {
        wrapper,
      }).container,
    ).toMatchSnapshot());

  test("should render with default size and custom viewbox", () => expect(
    render(
      <ArrowIcon viewBoxSize={32} />, {wrapper}
    ).container
  ).toMatchSnapshot())
});
