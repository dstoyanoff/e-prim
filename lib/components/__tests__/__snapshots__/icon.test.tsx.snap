// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`icon should render icon 1`] = `
.emotion-0 {
  width: 32px;
  height: 32px;
  color: #898896;
}

<div>
  <svg
    class="emotion-0"
    fill="none"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4 12H20M20 12L12 4M20 12L12 20"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
    />
  </svg>
</div>
`;

exports[`icon should render with default size and custom viewbox 1`] = `
.emotion-0 {
  width: 24px;
  height: 24px;
}

<div>
  <svg
    class="emotion-0"
    fill="none"
    viewBox="0 0 32 32"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M4 12H20M20 12L12 4M20 12L12 20"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
    />
  </svg>
</div>
`;
